File: apps/ios/Packages/StoryArcKit/Sources/Persistence/ReaderPreferences.swift
public import Foundation

public import StoryArcCore

public struct ReaderPreferences {
    private let defaults: UserDefaults
    /// The one fit the whole library used to share. Read once, folded into the
    /// fixed-layout default, and removed. See ``themes()``.
    private let legacyFitKey = "app.storyarc.pageFit"
    private let themesKey = "app.storyarc.themes"

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Every reading theme the reader has chosen, per shelf and per scope.
    ///
    /// One blob rather than a key per shelf: the whole point of `ShelfMemory` is that
    /// resolution walks from shelf to scope to built-in default, and a store that
    /// scattered the entries across `UserDefaults` keys would have to reimplement
    /// that walk. Unreadable stored data reads as no data — a theme is a preference,
    /// and losing one is worth far less than refusing to open the book.
    ///
    /// It is also where the page fit is picked up from where it used to live. The fit
    /// was one value for the whole library before `comic-reader`'s "persists per series"
    /// was honoured, and a reader who had chosen fit-to-width would otherwise find every
    /// comic they own back at fit-to-screen on the day they updated. So the old value
    /// becomes the fixed-layout *default*: every shelf that has not been told otherwise
    /// inherits it, which is exactly what "global" meant, and a shelf they set later
    /// keeps its own. The old key is removed as it is folded in, so this happens once.
    public func themes() -> ShelfMemory {
        let memory = storedThemes()
        guard let raw = defaults.string(forKey: legacyFitKey),
              let fit = PageFit(rawValue: raw)
        else { return memory }
        let migrated = memory.settingDefault(
            memory.default(for: .fixedLayout).settingFit(fit),
            for: .fixedLayout
        )
        defaults.removeObject(forKey: legacyFitKey)
        save(migrated)
        return migrated
    }

    public func save(_ memory: ShelfMemory) {
        guard let data = try? JSONEncoder().encode(memory) else { return }
        defaults.set(data, forKey: themesKey)
    }

    private func storedThemes() -> ShelfMemory {
        guard let data = defaults.data(forKey: themesKey),
              let memory = try? JSONDecoder().decode(ShelfMemory.self, from: data)
        else { return ShelfMemory() }
        return memory
    }
}

File: apps/android/core/persistence/src/main/kotlin/app/storyarc/core/persistence/ReaderPreferences.kt
package app.storyarc.core.persistence

import android.content.Context
import android.content.SharedPreferences
import app.storyarc.core.model.PageFit
import app.storyarc.core.model.ShelfMemory
import app.storyarc.core.model.ThemeScope
import kotlinx.serialization.json.Json

class ReaderPreferences(private val preferences: SharedPreferences) {

    companion object {
        fun open(context: Context): ReaderPreferences =
            ReaderPreferences(
                context.getSharedPreferences("app.storyarc.reader", Context.MODE_PRIVATE),
            )

        /**
         * The one fit the whole library used to share.
         *
         * Read once, folded into the fixed-layout default, and removed. See [themes].
         */
        private const val LEGACY_FIT = "pageFit"
        private const val THEMES = "themes"

        /**
         * Lenient about fields it does not know.
         *
         * A build that adds an axis must still read what an earlier build wrote, and
         * an older build reading a newer file should drop the field rather than the
         * whole theme.
         */
        private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    }

    /**
     * Every reading theme the reader has chosen, per shelf and per scope.
     *
     * One blob rather than a key per shelf: the whole point of [ShelfMemory] is that
     * resolution walks from shelf to scope to built-in default, and a store that
     * scattered the entries across preference keys would have to reimplement that
     * walk. Unreadable stored data reads as no data — a theme is a preference, and
     * losing one is worth far less than refusing to open the book.
     *
     * It is also where the page fit is picked up from where it used to live. The fit was
     * one value for the whole library before `comic-reader`'s "persists per series" was
     * honoured, and a reader who had chosen fit-to-width would otherwise find every comic
     * they own back at fit-to-screen on the day they updated. So the old value becomes the
     * fixed-layout *default*: every shelf that has not been told otherwise inherits it,
     * which is exactly what "global" meant, and a shelf they set later keeps its own. The
     * old key is removed as it is folded in, so this happens once.
     */
    fun themes(): ShelfMemory {
        val memory = storedThemes()
        val fit = preferences.getString(LEGACY_FIT, null)
            ?.let { name -> runCatching { PageFit.valueOf(name) }.getOrNull() }
            ?: return memory
        val migrated = memory.settingDefault(
            memory.default(ThemeScope.FIXED_LAYOUT).copy(fit = fit),
            ThemeScope.FIXED_LAYOUT,
        )
        preferences.edit().remove(LEGACY_FIT).apply()
        save(migrated)
        return migrated
    }

    fun save(memory: ShelfMemory) {
        preferences.edit().putString(THEMES, json.encodeToString(memory)).apply()
    }

    private fun storedThemes(): ShelfMemory =
        preferences.getString(THEMES, null)
            ?.let { runCatching { json.decodeFromString<ShelfMemory>(it) }.getOrNull() }
            ?: ShelfMemory()
}
